test/state/precompiles.cpp

```diff
@@ -3,11 +3,183 @@
 // SPDX-License-Identifier: Apache-2.0
 
 #include "precompiles.hpp"
+#include <intx/intx.hpp>
+#include <bit>
+#include <cassert>
+#include <iostream>
+#include <limits>
+#include <unordered_map>
 
 namespace evmone::state
 {
+using evmc::bytes;
+using evmc::bytes_view;
 using namespace evmc::literals;
 
+namespace
+{
+constexpr auto GasCostMax = std::numeric_limits<int64_t>::max();
+
+struct PrecompileAnalysis
+{
+    int64_t gas_cost;
+    size_t max_output_size;
+};
+
+inline constexpr int64_t num_words(size_t size_in_bytes) noexcept
+{
+    return static_cast<int64_t>((size_in_bytes + 31) / 32);
+}
+
+template <int BaseCost, int WordCost>
+inline constexpr int64_t cost_per_input_word(size_t input_size) noexcept
+{
+    return BaseCost + WordCost * num_words(input_size);
+}
+
+PrecompileAnalysis ecrecover_analyze(bytes_view /*input*/, evmc_revision /*rev*/) noexcept
+{
+    return {3000, 32};
+}
+
+PrecompileAnalysis sha256_analyze(bytes_view input, evmc_revision /*rev*/) noexcept
+{
+    return {cost_per_input_word<60, 12>(input.size()), 32};
+}
+
+PrecompileAnalysis ripemd160_analyze(bytes_view input, evmc_revision /*rev*/) noexcept
+{
+    return {cost_per_input_word<600, 120>(input.size()), 32};
+}
+
+PrecompileAnalysis identity_analyze(bytes_view input, evmc_revision /*rev*/) noexcept
+{
+    return {cost_per_input_word<15, 3>(input.size()), input.size()};
+}
+
+PrecompileAnalysis ecadd_analyze(bytes_view /*input*/, evmc_revision rev) noexcept
+{
+    return {rev >= EVMC_ISTANBUL ? 150 : 500, 64};
+}
+
+PrecompileAnalysis ecmul_analyze(bytes_view /*input*/, evmc_revision rev) noexcept
+{
+    return {rev >= EVMC_ISTANBUL ? 6000 : 40000, 64};
+}
+
+PrecompileAnalysis ecpairing_analyze(bytes_view input, evmc_revision rev) noexcept
+{
+    const auto base_cost = (rev < EVMC_ISTANBUL) ? 100000 : 45000;
```

Comment:
The other two use the  `rev >= ISTANBUL` order, perhaps nicer to make it consistent.